docs/spec/relayer/Definitions.md

```diff
@@ -0,0 +1,267 @@
+GetCurrentTimestamp(chainB) uint64
+
+// Verify that the data is written at the given path using provided membership proof and the root hash. 
+VerifyMembership(rootHash []byte, 
+                 proofHeight Height, 
+                 proof MembershipProof, 
+                 path String,
+                 data []byte) boolean
+
+// Create IBC datagram as part of processing event at chainA.
+CreateDatagram(ev IBCEvent, 
+               chainA Chain, 
+               chainB Chain, 
+               installedHeight Height) (IBCDatagram, Error)
+
+// Create UpdateClient datagrams from the list of signed headers
+CreateUpdateClientDatagrams(shs []SignedHeader) IBCDatagram[]
+
+// Submit given datagram to a given chain 
+Submit(chain Chain, datagram IBCDatagram) Error 
+
+// Return the correspondin chain for a given chainID 
+// We assume that the relayer maintains a map of known chainIDs and the corresponding chains.                
+GetChain(chainID String) Chain
+```
+
+For functions that return proof, if proof != nil, then the returned value is being verified. 
+The value is being verified using the header's app hash that is provided by the corresponding light client.
+
+### Error handling
+
+Helper functions listed above assume querying (parts of the) application state using Tendermint RPC. For example,
+`GetChannel` relies on `QueryChannel`. RPC calls can fail as 
+
+- no response is received within some timeout or
+- malformed response is received.
+
+In both cases, error handling logic should be defined by the caller. For example, in the former case, the caller might
+retry sending the same request to a same provider (full node), while in the latter case the request might be sent to 
+some other provider node. Although these kinds of errors could be due to network infrastructure issues, it is normally
+simpler to blame the provider (assume implicitly network is always correct and reliable). Therefore, correct provider
+always respond timely with a correct response, while in case of errors we consider the provider node faulty, and then 
+we replace it with a different node. 
+
+We now show the pseudocode for one of those functions that contains simplified error handling logic:
+
+```go
+func GetChannel(chain Chain, 
+           portId Identifier, 
+           channelId Identifier,  
+           proofHeight Height) (ChannelEnd, CommitmentProof) {
+
+    while(true) {
+        // Query provable store exposed by the full node of chain. 
+        // The path for the channel end is at channelEnds/ports/{portId}/channels/{channelId}".
+        // The membership proof returned is read at height proofHeight. 
+        channel, proof, error = QueryChannel(chain.provider, portId, channelId, proofHeight) 
+        if error != nil {
+            // elect a new provider from the peer list
+            if !ReplaceProvider(chain) { return (nil, nil) }  // return if fail to elect new provider         
```

Comment:
This is a bit different than the comment earlier where the caller replaces the provider. I guess it is a bit hard to figure the best way until we start coding this. But the higher lever function, `handleEvent()`, function could bubble up `drop`, `provider_error`, `retry`, etc. And the caller would be the one to take the action, like change the provider node, buffer the event for later retry, drop, etc.

Reply:
Yep; I realised I was not consistent with throwing errors and handling them. I will modify it to return error and then we can have handleEvent() caller dealing with it. We might want to sketch also that functionality (it's a bit like light client supervisor or fast sync scheduler) as it will need to implement error handling policies, store events that need to be retried, manage timeout triggers needed by packets handling, etc.